File: marlight_C#/Marlight_dll/Marlight.cs



using System;
using System.Threading;

namespace Marlight_lib
{
    /// <summary>
    /// Class for work with Marlight bulb thre WiFi receiver
    /// </summary>
    public class Marlight
    {
        /// <summary>
        /// valid IP address or host name
        /// </summary>
        public string IP = "localhost";

        /// <summary>
        /// port number on remote host
        /// </summary>
        public int Port = 50000;

        /// <summary>
        /// Dealy in milliseconds after any UDP command
        /// </summary>
        public int Delay = 0;

        /// <summary>
        /// Send message to remote host using threads
        /// If you use THREADS
        /// </summary>
        public bool UseThreads = true;

        /// <summary>
        /// Initilize library
        /// </summary>
        /// <param name="ip_host"></param>
        /// <param name="port"></param>
        /// <param name="delay"></param>
        /// <param name="use_threads"></param>
        public void Init(string ip_host, int port, int delay = 0, bool use_threads=true){
            IP = ip_host;
            Port = port;
            Delay = delay;
            UseThreads = use_threads;
        }

        #region Message Constats

        // ----------------------------- FIRST PAGE IPHONE APPLICATION  -----------------------------------
        readonly byte[] ALL_ON = new byte[] { 0x01, 0x55 };
        readonly byte[] ALL_OFF = new byte[] { 0x02, 0x55 };

        readonly byte[] BRIGHT_UP = new byte[] { 0x03, 0x55 };
        readonly byte[] BRIGHT_DOWN = new byte[] { 0x04, 0x55 };

        readonly byte[] TEMP_COLDER = new byte[] { 0x05, 0x55 };
        readonly byte[] TEMP_WARMER = new byte[] { 0x06, 0x55 };

        readonly byte[] BRIGHT_TEMP_DEFAULT = new byte[] { 0x07, 0x55 };

        readonly byte[] ON_1 = new byte[] { 0x08, 0x55 };
        readonly byte[] OFF_1 = new byte[] { 0x09, 0x55 };
        readonly byte[] ON_2 = new byte[] { 0x0A, 0x55 };
        readonly byte[] OFF_2 = new byte[] { 0x0B, 0x55 };
        readonly byte[] ON_3 = new byte[] { 0x0C, 0x55 };
        readonly byte[] OFF_3 = new byte[] { 0x0D, 0x55 };
        readonly byte[] ON_4 = new byte[] { 0x0E, 0x55 };
        readonly byte[] OFF_4 = new byte[] { 0x0F, 0x55 };

        // ----------------------------- SECOND PAGE IPHONE APPLICATION  (RGB) -----------------------------------
        readonly byte[] RGB_MODE_ON = new byte[] { 0x12, 0x55 };
        readonly byte[] RGB_MODE_BRIGHT_DOWN = new byte[] { 0x10, 0x55 };
        readonly byte[] RGB_MODE_BRIGHT_UP = new byte[] { 0x11, 0x55 };
        readonly byte[] RGB_MODE_SET_COLOR = new byte[] { 0x13, 0x00, 0x00, 0x00, 0x55 };

        // ----------------------------- THIRD PAGE IPHONE APPLICATION  (PRESETS) -----------------------------------
        readonly byte[] MODE_NIGHT = new byte[] { 0x14, 0x55 };
        readonly byte[] MODE_MEETING = new byte[] { 0x15, 0x55 };
        readonly byte[] MODE_READING = new byte[] { 0x16, 0x55 };
        readonly byte[] MODE_MODE = new byte[] { 0x17, 0x55 };                                        

        readonly byte[] MODE_TIMER = new byte[] { 0x18, 0x00, 0x00, 0x00, 0x00, 0x09, 0x14, 0x55 };    
        readonly byte[] MODE_ALARM = new byte[] { 0x19, 0x00, 0x00, 0x09, 0x14, 0x55 };                
        readonly byte[] MODE_SLEEP = new byte[] { 0x1A, 0x55 };                                        // ??????
        readonly byte[] MODE_RECREATION = new byte[] { 0x1B, 0x00, 0x00, 0x00, 0x55 };                  
        // ----------------------------- END -----------------------------------

        #endregion


        /// <summary>
        /// Send message to remote host
        /// </summary>
        /// <param name="message"></param>
        private void sendMsg(byte[] message)
        {

            if (UseThreads)
            {
                Thread t = new Thread(_sendMsg);
                t.Start(message);
            }
            else
            {
                _sendMsg(message);
            }
        }


        private void _sendMsg(object param)
        {
            byte[] message = (param as byte[]);
            using (System.Net.Sockets.UdpClient client = new System.Net.Sockets.UdpClient())
            {
                client.Connect(IP, Port);
                client.Send(message, message.Length);
                if (Delay != 0)
                    Thread.Sleep(Delay);
            }
        }


        /// <summary>
        /// Send custom message to remote host (for future use or undocument commands)
        /// </summary>
        /// <param name="custom_message"></param>
        public void SendCustomMsg(byte[] custom_message)
        {
            sendMsg(custom_message);
        }

        #region Simple commands
        /// <summary>
        /// Turn the lamp_num ON or make a lamp_num active by number
        /// Valid numbers is 1,2,3 and 4
        /// </summary>
        /// <param name="lamp_num"></param>
        public void LampOn(int lamp_num)
        {
            switch (lamp_num)
            {
                case 1:
                    sendMsg(ON_1);
                    break;
                case 2:
                    sendMsg(ON_2);
                    break;
                case 3:
                    sendMsg(ON_3);
                    break;
                case 4:
                    sendMsg(ON_4);
                    break;
                default:
                    break;
            }
        }

        /// <summary>
        /// Turn the lamp_num OFF
        /// Valid numbers is 1,2,3 and 4
        /// </summary>
        /// <param name="lamp_num"></param>
        public void LampOff(int lamp_num)
        {
            switch (lamp_num)
            {
                case 1:
                    sendMsg(OFF_1);
                    break;
                case 2:
                    sendMsg(OFF_2);
                    break;
                case 3:
                    sendMsg(OFF_3);
                    break;
                case 4:
                    sendMsg(OFF_4);
                    break;
                default:
                    break;
            }
        }

        /// <summary>
        /// Power on
        /// </summary>
        public void AllOn()
        {
            sendMsg(ALL_ON);
        }

        /// <summary>
        /// Power off
        /// </summary>
        public void AllOff()
        {
            sendMsg(ALL_OFF);
        }

        /// <summary>
        /// Bright Up on last lamp
        /// </summary>
        public void BrightUp()
        {
            sendMsg(BRIGHT_UP);
        }

        /// <summary>
        /// Bright Down on last lamp
        /// </summary>
        public void BrightDown()
        {
            sendMsg(BRIGHT_DOWN);
        }

        /// <summary>
        /// Set lamp temperature little bit colder
        /// </summary>
        public void TempColder()
        {
            sendMsg(TEMP_COLDER);
        }

        /// <summary>
        /// Set lamp temperature little bit warmer
        /// </summary>
        public void TempWarmer()
        {
            sendMsg(TEMP_WARMER);
        }

        /// <summary>
        /// Set the lamp to default color temperature and brightness
        /// </summary>
        public void SetDefaultTempAndBright()
        {
            sendMsg(BRIGHT_TEMP_DEFAULT);
        }
        #endregion

        #region RGB commands
        /// <summary>
        /// Set brightness little bit UP in RGB MODE
        /// </summary>
        public void RGBBrightUp()
        {
            sendMsg(RGB_MODE_BRIGHT_UP);
        }

        /// <summary>
        /// Set brightness little bit DOWN in RGB MODE
        /// </summary>
        public void RGBBrightDown()
        {
            sendMsg(RGB_MODE_BRIGHT_DOWN);
        }

        /// <summary>
        /// Turn ON RGB MODE
        /// </summary>
        public void RGBModeOn()
        {
            sendMsg(RGB_MODE_ON);
        }

        /// <summary>
        /// Set lamp color on RGB mode (RGB channels)
        /// </summary>
        /// <param name="R"> Red channel value</param>
        /// <param name="G"> Green channel value</param>
        /// <param name="B"> Blue channel value</param>
        public void RGBSetColor(byte R, byte G, byte B)
        {
            Byte[] color_msg = RGB_MODE_SET_COLOR;
            color_msg[1] = R;
            color_msg[2] = G;
            color_msg[3] = B;
            sendMsg(color_msg);
        }
        #endregion RGB commands

        #region Preset commands
        /// <summary>
        /// Preset a NIGHT MODE
        /// </summary>
        public void PresetNight()
        {
            sendMsg(MODE_NIGHT);
        }

        /// <summary>
        /// Preset a MEETING MODE
        /// </summary>
        public void PresetMeeting()
        {
            sendMsg(MODE_MEETING);
        }

        /// <summary>
        /// Preset a READING MODE
        /// </summary>
        public void PresetReading()
        {
            sendMsg(MODE_READING);
        }

        /// <summary>
        /// ????? just send a 2 byte to remote hoste (equal to MODE button oin iOS remote control application)
        /// </summary>
        public void PresetMode()
        {
            sendMsg(MODE_MODE);
        }

        /// <summary>
        /// Timer MODE
        /// </summary>
        /// <param name="on_hour"></param>
        /// <param name="on_minutes"></param>
        /// <param name="off_hour"></param>
        /// <param name="off_minutes"></param>
        public void PresetTimer(byte on_hour, byte on_minutes, byte off_hour, byte off_minutes)
        {
            byte[] timer_msg = MODE_TIMER;
            timer_msg[1] = on_hour;
            timer_msg[2] = on_minutes;
            timer_msg[3] = off_hour;
            timer_msg[4] = off_minutes;
            sendMsg(timer_msg);
        }

        /// <summary>
        /// emulate default send message on button TIMER on remote control on iOS
        /// </summary>
        public void PresetTimer()
        {
            sendMsg(MODE_TIMER);
        }

        /// <summary>
        /// Experemental ALARM MODE ???
        /// </summary>
        /// <param name="hours">??? hours ???</param>
        /// <param name="minutes">??? minutes ???</param>
        public void PresetAlarm(byte hours, byte minutes)
        {
            byte[] alarm_msg = MODE_ALARM;
            alarm_msg[1] = hours;
            alarm_msg[2] = minutes;
            sendMsg(alarm_msg);
        }

        /// <summary>
        /// emulate default send message on button SLEEP on remote control on iOS
        /// </summary>
        public void PresetSleep()
        {
            sendMsg(MODE_SLEEP);
        }

        /// <summary>
        /// Experemental RECREATION MODE ???
        /// </summary>
        /// <param name="byte_R">RGB red value</param>
        /// <param name="byte_G">RGB green value</param>
        /// <param name="byte_B">RGB blue value</param>
        public void PresetRecreation(byte byte_R, byte byte_G, byte byte_B)
        {
            byte[] recreation_msg = MODE_RECREATION;
            recreation_msg[1] = byte_R;
            recreation_msg[2] = byte_G;
            recreation_msg[3] = byte_B;
            sendMsg(recreation_msg);
        }
        #endregion Preset commands






    }
}

File: marlight_php/marlight.php
<?php

// ------ setup section (you can change it)

$server_ip = '192.168.1.80'; //remote host       (marlight wi-fi controller address)
$server_port = 50000; //remote host port  (marlight wi-fi controller port)

$delay = 0.1;         //delay after send a packet before send another (0.1 will be enaougth)

$error_value = 'Error'; //return message if error occurs
$ok_value = 'Ok'; //return message if packet send without error

$debug = true;                  //show debug info in response

$use_password = false; //use password for security
$password = '12345'; //password to compare with

// ------ end setup section











#region marlight udp packets const (do not change it)

// ----------------------------- FIRST PAGE IPHONE APPLICATION  ------------------------------------------
$ALL_ON                 = array(0x01, 0x55);
$ALL_OFF                = array(0x02, 0x55);

$BRIGHT_UP              = array(0x03, 0x55);
$BRIGHT_DOWN            = array(0x04, 0x55);

$TEMP_COLDER            = array(0x05, 0x55);
$TEMP_WARMER            = array(0x06, 0x55);

$BRIGHT_TEMP_DEFAULT    = array(0x07, 0x55);

$ON_1                   = array(0x08, 0x55);
$OFF_1                  = array(0x09, 0x55);
$ON_2                   = array(0x0A, 0x55);
$OFF_2                  = array(0x0B, 0x55);
$ON_3                   = array(0x0C, 0x55);
$OFF_3                  = array(0x0D, 0x55);
$ON_4                   = array(0x0E, 0x55);
$OFF_4                  = array(0x0F, 0x55);

// ----------------------------- SECOND PAGE IPHONE APPLICATION  (RGB) -----------------------------------
$RGB_MODE_ON            = array(0x12, 0x55);
$RGB_MODE_BRIGHT_DOWN   = array(0x10, 0x55);
$RGB_MODE_BRIGHT_UP     = array(0x11, 0x55);
$RGB_MODE_SET_COLOR     = array(0x13, 0x00, 0x00, 0x00, 0x55);

// ----------------------------- THIRD PAGE IPHONE APPLICATION  (PRESETS) --------------------------------
$MODE_NIGHT             = array(0x14, 0x55);
$MODE_MEETING           = array(0x15, 0x55);
$MODE_READING           = array(0x16, 0x55);
$MODE_MODE              = array(0x17, 0x55);

$MODE_TIMER             = array(0x18, 0x00, 0x00, 0x00, 0x00, 0x09, 0x14, 0x55);
$MODE_ALARM             = array(0x19, 0x00, 0x00, 0x09, 0x14, 0x55);
$MODE_SLEEP             = array(0x1A, 0x55); // ??????
$MODE_RECREATION        = array(0x1B, 0x00, 0x00, 0x00, 0x55);
// ----------------------------- END CONSTS -----------------------------------

#endregion


// function to send UDP packet
function sendMessage($udp_message)
{
    $message = vsprintf(str_repeat('%c', count($udp_message)), $udp_message);

    global $server_ip, $server_port, $delay, $debug, $error_value, $ok_value;

    if($debug){
        echo '<b>remote host/IP:</b> ' . $server_ip . '<br/>';
        echo '<b>remote port:</b> ' . $server_port . '<br/>';
        echo '<b>UDP message:</b> ';
        foreach($udp_message as $key => $value){
            echo '0x';
            printf("%02x\n", $value);
        }
        echo'<br/>';
        echo'<b>result:</b> ';
    }

    if ($socket = socket_create(AF_INET, SOCK_DGRAM, SOL_UDP)) {
        socket_sendto($socket, $message, strlen($message), 0, $server_ip, $server_port);
        sleep ($delay);
        socket_close($socket);

        echo $ok_value;
    } else {
        echo $error_value;
    }

    if($debug){
        echo '<br/>';
        echo '---------- debug info ----------<br/><br/>';
    }
}



//define is the POST or GET request
if ($_SERVER['REQUEST_METHOD'] === 'POST') {
    $request = $_POST;
}   else {
    $request = $_GET;
}


//if value set read them or set it to default
if(isset($request['command'])) {$command = $request['command'];} else {$command='';}

if(isset($request['r'])) {$r = $request['r'];} else {$r=0;}
if(isset($request['g'])) {$g = $request['g'];} else {$g=0;}
if(isset($request['b'])) {$b = $request['b'];} else {$b=0;}

if(isset($request['h1'])) {$h1 = $request['h1'];} else {$h1=0;}
if(isset($request['m1'])) {$m1 = $request['m1'];} else {$m1=0;}

if(isset($request['h2'])) {$h2 = $request['h2'];} else {$h2=0;}
if(isset($request['m2'])) {$m2 = $request['m2'];} else {$m2=0;}

if(isset($request['channel'])) {$channel = $request['channel'];} else {$channel=0;}

if(isset($request['password'])) {$pass = $request['password'];} else {$pass='';}


if($use_password && $pass!=$password) {
    echo 'You are dont autorized to this operation!';
    return;
}

// if degug mode
if($debug){
    echo '---------- debug info ----------<br/>';
    echo '<b>request method</b> '.$_SERVER['REQUEST_METHOD'].'<br/>';
    echo '<b>request value command:</b> '.$command.'<br/>';
    echo '<b>request value r:</b>  '.$r.'<br/>';
    echo '<b>request value g:</b>  '.$g.'<br/>';
    echo '<b>request value b:</b>  '.$b.'<br/>';

    echo '<b>request value h1:</b> '.$h1.'<br/>';
    echo '<b>request value m1:</b> '.$m1.'<br/>';
    echo '<b>request value h2:</b> '.$h2.'<br/>';
    echo '<b>request value m2:</b> '.$m2.'<br/>';
    if($channel!=0) {echo '<b>request command for channel:</b> '.$channel.'<br/>';}

}


// if request for specifyed channel then send ON command for this channel
if($channel!=0){
    $debug = false;
    switch($channel)        {
        case 1:
            sendMessage($ON_1); 
            break;
        case 2:
            sendMessage($ON_2); 
            break;
        case 3:
            sendMessage($ON_3); 
            break;
        case 4:
            sendMessage($ON_4); 
            break;
    }
    $debug = true;
} 


//send command depending POST/GET request
switch($command){
    case 'ALL_ON': 
        sendMessage($ALL_ON); 
        break;  
    case 'ALL_OFF': 
        sendMessage($ALL_OFF); 
        break;  
    case 'BRIGHT_UP': 
        sendMessage($BRIGHT_UP); 
        break;  
    case 'BRIGHT_DOWN': 
        sendMessage($BRIGHT_DOWN); 
        break;  
    case 'TEMP_COLDER': 
        sendMessage($TEMP_COLDER); 
        break;
    case 'TEMP_WARMER': 
        sendMessage($TEMP_WARMER);
        break;
    case 'BRIGHT_TEMP_DEFAULT': 
        sendMessage($BRIGHT_TEMP_DEFAULT);
        break;
    case 'ON_1': 
        sendMessage($ON_1); 
        break;
    case 'OFF_1': 
        sendMessage($OFF_1); 
        break;
    case 'ON_2':
        sendMessage($ON_2); 
        break;
    case 'OFF_2': 
        sendMessage($OFF_2); 
        break;
    case 'ON_3': 
        sendMessage($ON_3);
        break;
    case 'OFF_3':
        sendMessage($OFF_3);
        break;
    case 'ON_4': 
        sendMessage($ON_4);
        break;
    case 'OFF_4': 
        sendMessage($OFF_4); 
        break;
    case 'RGB_MODE_ON': 
        sendMessage($RGB_MODE_ON); 
        break;
    case 'RGB_MODE_BRIGHT_DOWN': 
        sendMessage($RGB_MODE_BRIGHT_DOWN);
        break;
    case 'RGB_MODE_BRIGHT_UP': 
        sendMessage($RGB_MODE_BRIGHT_UP); 
        break;
    case 'RGB_MODE_SET_COLOR':
        $rgb_msg =$RGB_MODE_SET_COLOR;
        $rgb_msg[1] = $r;
        $rgb_msg[2] = $g;
        $rgb_msg[3] = $b;
        sendMessage($rgb_msg);
        break;
    case 'MODE_NIGHT': 
        sendMessage($MODE_NIGHT); 
        break;
    case 'MODE_MEETING': 
        sendMessage($MODE_MEETING); 
        break;
    case 'MODE_READING': 
        sendMessage($MODE_READING); 
        break;
    case 'MODE_MODE': 
        sendMessage($MODE_MODE); 
        break;
    case 'MODE_TIMER': 
        $timer_msg =$MODE_TIMER;
        $timer_msg[1] = $h1;
        $timer_msg[2] = $m1;
        $timer_msg[3] = $h2;
        $timer_msg[4] = $m2;
        sendMessage($timer_msg);
        break;
    case 'MODE_ALARM': 
        $alarm_msg =$MODE_ALARM;
        $alarm_msg[1] = $h1;
        $alarm_msg[2] = $m1;
        sendMessage($alarm_msg);
        break;
    case 'MODE_SLEEP': 
        sendMessage($MODE_SLEEP); 
        break;
    case 'MODE_RECREATION': 
        $recreation_msg =$MODE_RECREATION;
        $recreation_msg[1] = $R;
        $recreation_msg[2] = $G;
        $recreation_msg[3] = $B;
        sendMessage($recreation_msg);
        break;
    default:
        break;
}